src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -45,6 +45,9 @@
 
     private int currentIndex = -1;
 
+    // Don't re-resolve on first next() call
+    private boolean connectedSinceNext = true;
```

Comment:
My reasoning was only defensively, but on a 2nd look, it would be overkill. So, no problem without being volatile, imo.